src/runtime-rs/crates/hypervisor/README.md

```diff
@@ -0,0 +1,79 @@
+This function initialize the runtime handler. It will register the plugins into the HYPERVISOR_PLUGINS. Different plugins are needed for different hypervisors. 
+
+This link is for QEMU. Other VMM plugin methods haven't support currently.
+[https://github.com/kata-containers/kata-containers/blob/main/src/libs/kata-types/src/config/hypervisor/qemu.rs](https://github.com/kata-containers/kata-containers/blob/main/src/libs/kata-types/src/config/hypervisor/qemu.rs)
+qemu plugin defines the methods to adjust and validate the hypervisor config file, those methods could be modified if it is needed.
+
+[https://github.com/kata-containers/kata-containers/blob/main/src/runtime-rs/crates/runtimes/src/manager.rs#L107](https://github.com/kata-containers/kata-containers/blob/main/src/runtime-rs/crates/runtimes/src/manager.rs#L107)
+After that, when loading the toml config, the plugins will be called to adjust and validate config file.
+
+### new_instance
+This function will create a runtime_instance which include the operations for container and sandbox.  At the same time, a hypervisor instance will be created.  Qemu instance will be created here as well, and set the hypervisor config file
+[https://github.com/kata-containers/kata-containers/blob/main/src/runtime-rs/crates/runtimes/virt_container/src/lib.rs#L101](https://github.com/kata-containers/kata-containers/blob/main/src/runtime-rs/crates/runtimes/virt_container/src/lib.rs#L101)
+
+## 2. Hypervisor Trait
+To support multi-vmm, the hypervisor trait need to be implemented. 
+[https://github.com/kata-containers/kata-containers/blob/main/src/runtime-rs/crates/hypervisor/src/lib.rs](https://github.com/kata-containers/kata-containers/blob/main/src/runtime-rs/crates/hypervisor/src/lib.rs)
+```rust
+pub trait Hypervisor: Send + Sync {
+    // vm manager
+    async fn prepare_vm(&self, id: &str, netns: Option<String>) -> Result<()>;
+    async fn start_vm(&self, timeout: i32) -> Result<()>;
+    async fn stop_vm(&self) -> Result<()>;
+    async fn pause_vm(&self) -> Result<()>;
+    async fn save_vm(&self) -> Result<()>;
+    async fn resume_vm(&self) -> Result<()>;
+    
+    // device manager
+    async fn add_device(&self, device: device::Device) -> Result<()>;
+    async fn remove_device(&self, device: device::Device) -> Result<()>;
+    
+    // utils
+    async fn get_agent_socket(&self) -> Result<String>;
+    async fn disconnect(&self);
+    async fn hypervisor_config(&self) -> HypervisorConfig;
+    async fn get_thread_ids(&self) -> Result<VcpuThreadIds>;
+    async fn get_pids(&self) -> Result<Vec<u32>>;
+    async fn cleanup(&self) -> Result<()>;
+    async fn check(&self) -> Result<()>;
+    async fn get_jailer_root(&self) -> Result<String>;
+    async fn save_state(&self) -> Result<HypervisorState>;
+   }
+```
+In current design, VM will be started in the following steps.
+
+![vmm start](../../docs/images/vm-start.svg)
+
+## 3. Directory Tree
```

Comment:
The directory tree is not very useful, and it's easy to be out-of-date. So is it better to delete it?

Reply:
fixed